portal/client-services/org.wso2.is.portal.user.client.api/src/main/java/org/wso2/is/portal/user/client/api/IdentityStoreClientServiceImpl.java

```diff
@@ -341,24 +353,136 @@ public String getPrimaryDomainName() throws UserPortalUIException {
+
+        List<UserListBean> userList;
+        List<User> users;
+
+        if (length < 0) {
+            length = MAX_RECORD_LENGTH;
+        }
+
+        if (StringUtils.isNullOrEmpty(domainName)) {
+            domainName = getPrimaryDomainName();
+        }
+
+        if (StringUtils.isNullOrEmpty(claimURI)
+                || StringUtils.isNullOrEmpty(claimValue)) {
+            return getUserList(offset, length, domainName, requestedClaims);
+        } else {
+            MetaClaim metaClaim = new MetaClaim();
+            metaClaim.setClaimUri(claimURI);
+            try {
+                users = getRealmService().getIdentityStore().listUsers(
+                        metaClaim, claimValue, offset, length, domainName);
+            } catch (IdentityStoreException e) {
+                String error = "Error while retrieving users for " + claimURI + "= " + claimValue;
+                LOGGER.error(error, e);
+                throw new UserPortalUIException(error);
+            }
+        }
+
+        userList = generateUserListBean(users, requestedClaims);
+
+        return userList;
+    }
+
+    @Override
+    public List<UserListBean> getUserList(int offset, int length, String domainName,
+                                          List<ClaimConfigEntry> requestedClaims) throws UserPortalUIException {
+        List<User> users;
+        if (length < 0) {
+            length = MAX_RECORD_LENGTH;
+        }
+        try {
+            users = getRealmService().getIdentityStore().listUsers(offset, length, domainName);
+        } catch (IdentityStoreException e) {
+            String error = "Error while retrieving users";
+            LOGGER.error(error, e);
+            throw new UserPortalUIException(error);
+        }
+
+        return generateUserListBean(users, requestedClaims);
+    }
 
     private RealmService getRealmService() {
         if (this.realmService == null) {
             throw new IllegalStateException("Realm Service is null.");
         }
         return this.realmService;
     }
+
+    private List<UserListBean> generateUserListBean(List<User> users, List<ClaimConfigEntry> requestedClaims)
+            throws UserPortalUIException {
+        List<UserListBean> userList = new ArrayList<>();
+
+        Map<String, String> claimDisplayNames = requestedClaims.stream()
+                        .collect(Collectors.toMap(ClaimConfigEntry::getClaimURI, ClaimConfigEntry::getDisplayName));
+
+        List<MetaClaim> metaClaims = requestedClaims.stream()
+                .map(claimConfigEntry -> new MetaClaim("", claimConfigEntry.getClaimURI()))
+                .collect(Collectors.toList());
+
+
+        List<MetaClaim> groupMetaClaims = new ArrayList<>();
+        MetaClaim groupMetaClaim = new MetaClaim();
+        groupMetaClaim.setClaimUri(GROUPNAME_CLAIM);
+        groupMetaClaims.add(groupMetaClaim);
+
+        for (User user : users) {
```

Comment:
Better to review this during a code review. There's a lot of API invokes which includes multiple db calls. This can have an effect on UX.